src/Transform/ONNX/InstrumentONNXPass.cpp

```diff
@@ -88,32 +88,34 @@ class InstrumentONNXPass
     if (instrumentONNXOps == "" || instrumentONNXOps == "NONE")
       return;
     init(instrumentONNXOps);
-    auto function = getFunction();
-    auto &funcBody = function.getBody();
 
-    // Iterate on the operations
-    for (Operation &op : funcBody.getOps()) {
-      if (isa<mlir::ONNXOpsDialect>(op.getDialect())) {
+    // Iterate on the operations nested in this function
```

Comment:
Question: does the walk works recursively over structured constructs like onnx.loop and onnx.if? Also, it walks in a tree and at the same time add operation. Out of curiosity, are they encountered in the walk (as added ops are also part of the function), but ignored as only ONNX ops are instrumented, and the added ops are KRNL ops. Can you confirm, thanks.

Nit, Doru likes period at the end of sentences.

Reply:
Walk() traverses the mlir IR structure recursively, according to the document. I tested and found the inserted operations were traversed no matter they were insert before or after the current node. The order seems to be pre-order, not specified in document.

Reply:
Thanks, looks like a very useful and robust approach to traversal